libraries/joomla/application/route.php

```
@@ -99,4 +107,29 @@ public static function _($url, $xhtml = true, $ssl = null)
 
 		return $url;
 	}
+
+	/**
+	 * Magic method to provide short access to the clients' route functions "site" and "administrator".
+	 * - To build a route for site:          <var>JRoute::site($url)</var>.
+	 * - To build a route for administrator: <var>JRoute::administrator($url)</var>
+	 *
+	 * @param   string  $name       The called method name
+	 * @param   array   $arguments  The method arguments
+	 *
+	 * @return  string
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public static function __callStatic($name, $arguments)
+	{
+		if (count($arguments))
+		{
+			$xhtml = isset($arguments[1]) ? $arguments[1] : true;
+			$ssl   = isset($arguments[2]) ? $arguments[2] : null;
+
+			return static::_($arguments[0], $xhtml, $ssl, $name);
+		}
+
+		return null;
```

This should be a `trigger_error()` call because you in essence have a "call to undefined function" error here.

---

If we check for a valid client name here, the behavior would be different when called using undescore function. That function fails silently on invalid router and we need to keep it like that for B/C, imho.

And here we are only checking for missing first parameter coz we need to populate 4th parameter.

Another alternate way that comes to my mind is to not modify the undescore function signature and add another common function that accepts client name as first parameter.

We might then not need the magic function too. Your opinion please.

---

Personally I hate the use of magic functions.  They usually expose unwanted behavior and do not handle error conditions appropriately (in an error state they should do the same thing as the native PHP engine does for whatever that magic function is overloading).  Even in the Framework's database repo, I removed `__call()` and just made "real" stub methods for the shorthand `q()`, `qn()`, and `e()` methods.

So personally I'd just add stub real methods and avoid the magic as a first option.  If the magic method is really needed, then it needs to error trap/handle properly.

> That function fails silently on invalid router and we need to keep it like that for B/C, imho.

If the system can't generate a route, it shouldn't silently fail over.  There needs to be an error raised.